components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/SAMLAssertionClaimsCallback.java

```diff
@@ -57,7 +84,211 @@ public void handleCustomClaims(IDTokenBuilder builder, OAuthTokenReqMessageConte
             } else {
                 log.debug("No AttributeStatement found! ");
             }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Adding claims for user " + requestMsgCtx.getAuthorizedUser() + " to id token.");
+            }
+            try {
+                JSONArray values;
+                Map<String, Object> claims = getResponse(requestMsgCtx);
+                Object claimSeparator = claims.get(MULTI_ATTRIBUTE_SEPARATOR);
+                if (claimSeparator != null) {
+                    userAttributeSeparator = (String) claimSeparator;
+                    claims.remove(MULTI_ATTRIBUTE_SEPARATOR);
+                }
+
+                for (Map.Entry<String, Object> entry : claims.entrySet()) {
+                    String value = entry.getValue().toString();
+                    values = new JSONArray();
+                    if (userAttributeSeparator != null && value.contains(userAttributeSeparator)) {
+                        StringTokenizer st = new StringTokenizer(value, userAttributeSeparator);
+                        while (st.hasMoreElements()) {
+                            String attValue = st.nextElement().toString();
+                            if (attValue != null && attValue.trim().length() > 0) {
+                                values.add(attValue);
+                            }
+                        }
+                    } else {
+                        values.add(value);
+                    }
+                    builder.setClaim(entry.getKey(), values.toJSONString());
+                }
+            } catch (OAuthSystemException e) {
+                log.error(
+                        "Error occurred while adding claims of " + requestMsgCtx.getAuthorizedUser() + " to id token.",
+                        e);
+            }
+        }
+    }
+
+    /**
+     * Get response map
+     *
+     * @param requestMsgCtx Token request message context
+     * @return Mapped claimed
+     * @throws OAuthSystemException
+     */
+    private Map<String, Object> getResponse(OAuthTokenReqMessageContext requestMsgCtx)
+            throws OAuthSystemException {
+
+        Map<ClaimMapping, String> userAttributes =
+                getUserAttributesFromCache(requestMsgCtx.getProperty("accessToken").toString());
+        Map<String, Object> claims = Collections.EMPTY_MAP;
+
+        // If subject claim uri is null, we get the actual user name of the logged in user.
+        if ((userAttributes == null || userAttributes.isEmpty()) && (getSubjectClaimUri(requestMsgCtx) == null)) {
+            if (log.isDebugEnabled()) {
+                log.debug("User attributes not found in cache. Trying to retrieve attribute for user " + requestMsgCtx
+                        .getAuthorizedUser());
+            }
+            try {
+                claims = getClaimsFromUserStore(requestMsgCtx);
+            } catch (Exception e) {
```

Comment:
getClaimsFromUserStore() method was throwing Exception. Changed it to throw the specific exceptions and caught the specific exceptions here. https://github.com/wso2/carbon-identity/compare/master...nuwandi-is:refac#diff-f63f68423fee8790a53aabc15919f349R152